### ISSUE TITLE
"Error finalizing cipher loop" when running openssl speed -evp

### ISSUE THREAD

**Issue by tom-cosgrove-arm** [openssl/openssl#20291]
Since commit 98283a61f5 I am seeing

```
$ perl Configure darwin64-arm64 no-shared no-threads no-async -Wimplicit-function-declaration && time make -j10 && echo && DYLD_LIBRARY_PATH=. apps/openssl speed -evp aes-128-gcm
:
		apps/libapps.a -lssl -lcrypto  

real	0m34.00s
user	2m56.95s
sys	0m43.71s

Doing AES-128-GCM for 3s on 16 size blocks: Error finalizing cipher loop
197335957 AES-128-GCM's in 3.00s
Doing AES-128-GCM for 3s on 64 size blocks: Error finalizing cipher loop
135033964 AES-128-GCM's in 3.00s
Doing AES-128-GCM for 3s on 256 size blocks: Error finalizing cipher loop
53776484 AES-128-GCM's in 3.00s
Doing AES-128-GCM for 3s on 1024 size blocks: Error finalizing cipher loop
21636816 AES-128-GCM's in 3.00s
Doing AES-128-GCM for 3s on 8192 size blocks: Error finalizing cipher loop
3051110 AES-128-GCM's in 2.99s
Doing AES-128-GCM for 3s on 16384 size blocks: Error finalizing cipher loop
1537926 AES-128-GCM's in 3.00s
version: 3.2.0-dev
built on: Wed Feb 15 00:10:38 2023 UTC
options: bn(64,64)
compiler: cc -fPIC -arch arm64 -O3 -Wall -Wimplicit-function-declaration -DL_ENDIAN -DOPENSSL_PIC -DOPENSSL_BUILDING_OPENSSL -DNDEBUG
CPUINFO: OPENSSL_armcap=0x187d
The 'numbers' are in 1000s of bytes per second processed.
type             16 bytes     64 bytes    256 bytes   1024 bytes   8192 bytes  16384 bytes
AES-128-GCM    1052458.44k  2880724.57k  4588926.63k  7385366.53k  8359429.14k  8399126.53k
```

If I revert `98283a61f5` and build, I get

```
$ perl Configure darwin64-arm64 no-shared no-threads no-async -Wimplicit-function-declaration && time make -j10 && echo && DYLD_LIBRARY_PATH=. apps/openssl speed -evp aes-128-gcm
:
		apps/libapps.a -lssl -lcrypto  

real	0m33.77s
user	2m55.24s
sys	0m43.58s

Doing AES-128-GCM for 3s on 16 size blocks: 198213155 AES-128-GCM's in 3.00s
Doing AES-128-GCM for 3s on 64 size blocks: 136294089 AES-128-GCM's in 3.00s
Doing AES-128-GCM for 3s on 256 size blocks: 53526662 AES-128-GCM's in 3.00s
Doing AES-128-GCM for 3s on 1024 size blocks: 20822216 AES-128-GCM's in 2.96s
Doing AES-128-GCM for 3s on 8192 size blocks: 2997378 AES-128-GCM's in 2.97s
Doing AES-128-GCM for 3s on 16384 size blocks: 1535425 AES-128-GCM's in 3.00s
version: 3.2.0-dev
built on: Wed Feb 15 00:15:13 2023 UTC
options: bn(64,64)
compiler: cc -fPIC -arch arm64 -O3 -Wall -Wimplicit-function-declaration -DL_ENDIAN -DOPENSSL_PIC -DOPENSSL_BUILDING_OPENSSL -DNDEBUG
CPUINFO: OPENSSL_armcap=0x187d
The 'numbers' are in 1000s of bytes per second processed.
type             16 bytes     64 bytes    256 bytes   1024 bytes   8192 bytes  16384 bytes
AES-128-GCM    1057136.83k  2907607.23k  4567608.49k  7203361.21k  8267515.35k  8385467.73k
```

Part of the changes in that commit are

```
     if (decrypt)
-        EVP_DecryptFinal_ex(ctx, buf, &outl);
+        rc = EVP_DecryptFinal_ex(ctx, buf, &outl);
     else
-        EVP_EncryptFinal_ex(ctx, buf, &outl);
+        rc = EVP_EncryptFinal_ex(ctx, buf, &outl);
+
+    if (rc <= 1)
+        BIO_printf(bio_err, "Error finalizing cipher loop\n");
     return count;
 }
```

but `man EVP_DecryptFinal_ex` says

```
EVP_DecryptFinal_ex() returns 0 if the decrypt failed or 1 for success.
```

So: is the test for `rc <= 1` correct?

(I'm raising this as an issue because I think I've seen a number of changes recently regarding return values, and I haven't checked them all: this one was literally in my face!)

**Comment by mattcaswell** [comment]
Looks like a bug to me. Probably should be `<= 0`

**Comment by mattcaswell** [comment]
...or maybe just `rc == 0`

**Comment by beldmit** [comment]
Thanks! Fix in #20306 20306
